File: iter1/199.cpp
//
// Created by chenyufan on 2020/5/8.
//

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */


#include <vector>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

using namespace std;

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        vector<int> ret;
        _rightSideView(root, ret, 0);
        return ret;
    }

    void _rightSideView(TreeNode* root, vector<int>& pre, int curLevel) {
        if (!root) return;
        if (curLevel == pre.size()) pre.push_back(root->val);
        _rightSideView(root->right, pre, curLevel+1);
        _rightSideView(root->left, pre, curLevel+1);
    }

};

int main() {
    Solution s;
}

File: iter1/199.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
#     1            <---
#  /   \
# 2     3         <---
#  \     \
#   5     4       <---
#  /
# 6
#
# 1, 3, 4, 6

class Solution:
    def rightSideView(self, root: TreeNode) -> List[int]:
        ans = []
        max_level = -1
        def dfs(node, level):
            if not node: return
            if level > max_level:
                ans.append(node.val)
                max_level = level
            dfs(node.right, level+1)
            dfs(node.left, level+1)
        dfs(root, 0)
        return ans